Add a managed Git implementation

qmfrederik opened the pull request:
Adds a managed Git implementation to NBGV. The managed implementation is used by default but libgit2sharp is used under a switch or when _writing_ to git is required (e.g. `nbgv prepare-release`). 
 
All tests run against both backends.

This implementation already contains a couple of performance optimizations which I worked on with the help of @filipnavara ; but the performance may have regressed as I focused on matching the NBGV features.

Perhaps worth mentioning is that:
- Opening files (i.e. `File.OpenRead`, but also `File.Exists` and just constructing the file name) turned out to be an expensive operation. On Windows, there's for example a P/Invoke call to CreateFile to implement some kind of `TryOpen`, to avoid having to call `File.Exists` followed by `File.OpenRead` (which is slow and racy)
- I did not know about git pack files before starting this work, and a fair amount of the code is related to parsing git pack files in a performant manner (e.g. pooling `FileStream` objects which access git pack files, and caching objects retrieved from a git pack in a cache, because this can be an expensive operation).
- For git pack files, `Streams` which point to the pack file are pooled instead of opening/closing them repeatedly (which is expensive).

Notable steps completed:
- [X] Add unit tests for the managed Git client
- [X] Review exceptions being thrown, make them more descriptive
- [x] `GitCommit`: Since most commits only have one/two parents, investigate whether it is worth storing the first (two) parents as a field on the `GitCommit` struct, and additional parents in a list. This would avoid allocating many `List<GitObjectId>` objects with only a single item.
- [x] `GitCommit`: Investigate whether it's worth parsing the `Author` value for the current HEAD only.
- [X] Add documentation for the managed Git client
- [x] Add support for Git alternates
- [x] ~~Investigate whether walking the Git tree can be parallelized~~ (deferred)
- [x] Test against [very large pack files](https://github.com/libgit2/libgit2sharp/issues/1835).
- [x] Test with shallow clone and fail with a better exception message than libgit2sharp fails with (an NRE)
- [x] ~~Test that .ps1 scripts in nuget package still work~~ (these scripts were removed as they are incompatible now)
- [x] Implement `ManagedGit.GitRepository.ShortenObjectId`
- [x] Implement `GitRepository.Lookup`

Closes #505
Closes #487
Closes #433
Closes #92

Improves perf as tracked by #114

qmfrederik commented:
@AArnott let me know how you want to proceed with this - I'm opening the PR to get early feedback. I guess could first work to get the PR to a state where it's good to merge (e.g. code quality, unit tests,...), and then stabilize further (e.g. perf work,...) on a feature/version branch?

qmfrederik commented:
I ran benchmarks on this PR, and here are the results:

|            Method |             TestData |      Mean |     Error |    StdDev | Ratio | RatioSD |
|------------------ |--------------------- |----------:|----------:|----------:|------:|--------:|
| GetVersionLibGit2 |  Cuemon;version.json |  6.318 ms | 0.0821 ms | 0.0641 ms |  1.00 |    0.00 |
| GetVersionManaged |  Cuemon;version.json |  3.188 ms | 0.0667 ms | 0.0742 ms |  0.51 |    0.01 |
|                   |                      |           |           |           |       |         |
| GetVersionLibGit2 | NerdB(...).json [35] | 23.091 ms | 0.3804 ms | 0.3558 ms |  1.00 |    0.00 |
| GetVersionManaged | NerdB(...).json [35] |  8.400 ms | 0.1615 ms | 0.1728 ms |  0.36 |    0.01 |
|                   |                      |           |           |           |       |         |
| GetVersionLibGit2 | Super(...).json [24] | 25.678 ms | 0.5559 ms | 0.6827 ms |  1.00 |    0.00 |
| GetVersionManaged | Super(...).json [24] |  3.179 ms | 0.0599 ms | 0.0531 ms |  0.12 |    0.00 |
|                   |                      |           |           |           |       |         |
| GetVersionLibGit2 |   xunit;version.json | 25.977 ms | 0.5188 ms | 0.6176 ms |  1.00 |    0.00 |
| GetVersionManaged |   xunit;version.json | 16.945 ms | 0.3271 ms | 0.3636 ms |  0.65 |    0.02 |

qmfrederik commented:
@AArnott I think this PR is close to ready for review:
- VersionOracle and BuildIntegration tests pass
- Implemented known missing features (e.g. Git alternates)
- Added unit tests for newly added code

One item pending your feedback is the TargetFramework - using `Span<byte>`-based overloads in classes like `Stream` requires targetting `netstandard2.1`/`netcoreapp3.0` _or_ polyfill code. What's the lowest TargetFramework you want to support?

Could you have a first glance at this PR and let me know what you think?

AArnott commented:
I would prefer to multitarget in order to continue to support netcoreapp2.1/net472 while optimizing perf on netcoreapp3.1 through overrides of Span overloads.

I'll take a look at this PR later today.

Have you reviewed the various issues where we've discussed this for corner cases in git storage patterns? I can't recall where I saw it, but the libgit2sharp maintainers were warning us that there were many different cases in how packs/objects were stored in the `.git` folder and we would need to support _all_ of them or else nb.gv may fail at arbitrary times in a repo when git happened to pack in that particular way.

AArnott commented on src/NerdBank.GitVersioning.Tests/NerdBank.GitVersioning.Tests.csproj:
```diff
@@ -46,8 +46,41 @@
+    </None>
+    <None Update="Managed\tree.delta">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
```
Can we create a subdirectory for all such files and then a single msbuild item with wildcard to arrange for copying them all?

FWIW: I find across test runners that some don't always make finding these files easy, even if they are built to be next to the test .dll. In those cases I _embed_ the files instead of copying them to the output directory. So just know I'm ok with that option if you hit similar issues.

qmfrederik replied:
Yes, I ended up embedding the test files. For reasons I don't understand, `dotnet test` behaviors different on the Azure Agents than locally, _even_ with identical .NET Core versions. I tried running CI in docker containers (to avoid [customizations like this](https://github.com/actions/virtual-environments/blob/main/images/win/scripts/Installers/Install-DotnetSDK.ps1#L37-L41)), but there doesn't seem to be a read-to-use Docker image which has both net461 and netcoreapp2.1 available.

AArnott commented on src/NerdBank.GitVersioning/Managed/GitObjectId.cs:
```diff
@@ -0,0 +1,257 @@
+    /// object.
+    /// </summary>
+    /// <seealso href="https://git-scm.com/book/en/v2/Git-Internals-Git-Objects"/>.
+    public unsafe struct GitObjectId : IEquatable<GitObjectId>
```
In general, I'd prefer *all* these new types to be `internal` if at all possible. I don't want to support folks who use NB.GV as a managed git implementation for arbitrary stuff.

qmfrederik replied:
That's fine, if we can expose the internals to the test project.

AArnott replied:
No, please don't. `InternalsVisibleTo` opens the doors to testing all the wrong stuff. Sort of like the case you made about changing our tests from `GitExtensions` to `VersionOracle`. 
If tests really need to access this type, leave it `public`.

qmfrederik replied:
There's about 47 unit tests (all in the `NerdBank.GitVersioning.Tests.Managed` namespace) which directly access the Git-related classes. I found the unit tests to be useful; they revealed a couple of bugs. I'd vote for keeping them, if possible.

AArnott commented on src/NerdBank.GitVersioning/Managed/GitObjectId.cs:
```diff
@@ -0,0 +1,257 @@
+            Debug.Assert(value.Length == 20);
+
+            GitObjectId objectId = new GitObjectId();
+            Span<byte> bytes = new Span<byte>(objectId.value, NativeSize);
```
I wonder if we can arrange to run all the tests on an ARM machine in particular to help test for proper handling of endian differences. Thoughts?

qmfrederik replied:
Travis and Drone CI offer free-of-charge CI which runs on ARM VMs

AArnott replied:
Well, maybe adding a Travis CI is worth it. What do you think?
In the meantime, I've filed [a request for ARM agents](https://developercommunity.visualstudio.com/idea/1234435/offer-arm-hosted-agents-in-azure-pipelines.html). Feel free to vote it up.

AArnott commented on src/NerdBank.GitVersioning/Managed/GitObjectId.cs:
```diff
@@ -0,0 +1,257 @@
+        /// <remarks>
+        /// This method is used to populate file paths as byte* objects which are passed to Unicode-based
+        /// Windows APIs.</remarks>
+        public void CopyToUnicodeString(int start, int length, Span<byte> bytes)
```
The method name and all the docs talk about a "Unicode" string, but Unicode is comprised of many encodings. Can we be more precise here and in the docs? Calling it `Utf16` in the method name and specifying `UTF-16` in the docs everywhere, for example, if I'm reading the code right you're doing UTF-16.

qmfrederik replied:
Sure, I renamed the methods to use `Utf16`.

AArnott commented on src/NerdBank.GitVersioning/Managed/GitObjectId.cs:
```diff
@@ -0,0 +1,257 @@
+        {
+            fixed (byte* thisValue = this.value)
+            {
+                new Span<byte>(thisValue, NativeSize).CopyTo(value);
```
Does something prevent us from defining a private property on this struct that returns this `new Span<byte>` so we don't have to repeatedly create it? I know `Span<T>` comes with some syntax restrictions, so maybe that's why you're doing it repeatedly throughout this file, but I thought I'd check.

filipnavara replied:
Span can only exist on stack and it's essentially no-cost to create it with `new` since it's not really doing any allocation.

AArnott replied:
Yes, understood. It's just ease of reading that I was going for here. Is it legal C# to define and use this?

```cs
private Span<byte> ThisValue => new Span<byte>(this.value, NativeSize);
```

filipnavara replied:
(Span -> ReadOnlySpan here?)

qmfrederik replied:
```csharp
private Span<byte> ThisValue => new Span<byte>(this.value, NativeSize);
```

is not valid:

> **CS1666** You cannot use fixed size buffers contained in unfixed expressions. Try using the fixed statement.

AArnott replied:
Ah, that's easy to fix. I pushed [a simplification](https://github.com/dotnet/Nerdbank.GitVersioning/pull/521/commits/745d0d5bf3e07650532c18eeda12525fc32747e3) to your branch. Now `NativeSize` is only used once and everyone else uses the Span.

qmfrederik replied:
I don't think that fix is valid, though:

```csharp
        private Span<byte> Value
        {
            get
            {
                fixed (byte* value = this.value)
                {
                    return new Span<byte>(value, NativeSize);
                }
            }
        }
```

You'll return a `Span<byte>` which refers to the memory at `value`. But `value` is only `fixed` within the scope of the getter. If the CLR decides to move `GitObjectId` to another location, so will `this.value`, and the `Span` will refer to an invalid location, right?

AArnott replied:
It's generally a good concern, yes. But I'm pretty certain in this case it's fine.

The CLR has special treatment for Spans (which is why it is a  `ref struct` and only allowed on the stack -- not the heap). You don't even need a `fixed` block to get a span of a string, and yet we never have to worry about the GC moving the string's memory and invalidating the span. A span into any object effectively *pins* the object so the GC can't move it.
See https://adamsitnik.com/Span/#how-does-it-work for more info.

Now if we consider that this array is *not* a managed object at all, but rather an inline fixed sized array in a struct, we come up with another reason to not worry: The GC will *never* rearrange memory on the stack itself. It only rearranges memory on the heap. Since `GitObjectId` is a struct, any method executing within this type means the struct's data is on the stack (even if it had previously been `box`'d to the heap).

AArnott commented on src/NerdBank.GitVersioning/Managed/ManagedVersionOracle.cs:
```diff
@@ -0,0 +1,139 @@
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VersionOracle"/> class.
+        /// </summary>
+        internal ManagedVersionOracle(string projectDirectory, GitRepository repo, GitCommit? head, ICloudBuild cloudBuild, int? overrideVersionHeightOffset = null, string projectPathRelativeToGitRepoRoot = null)
```
I see a *lot* of code duplication between this and the `LibGit2VersionOracle` constructor. What can we do to consolidate?
Is it that although the code *looks* the same, it must compile against distinct APIs and thus must be written twice?

AArnott commented on src/NerdBank.GitVersioning/Managed/FileHelpers.cs:
```diff
@@ -0,0 +1,101 @@
+{
+    internal static class FileHelpers
+    {
+        [DllImport("kernel32.dll", EntryPoint = "CreateFileW", SetLastError = true, CharSet = CharSet.Unicode, BestFitMapping = false, ExactSpelling = true)]
```
I'd prefer to get these by depending on the `PInvoke.Kernel32` package if we can.

qmfrederik replied:
`Kernel32.CreateFile` is missing an overload which allows you to pass `filename` as `byte*`; `filename` is passed as `byte*` to avoid having to create `string` objects for the various paths.

AArnott replied:
eww.. The PInvoke project is supposed to prefer pointers to the managed types for exactly that reason. I'll get that fixed. But continue as you were in the meantime. :)

AArnott replied:
@qmfrederik https://github.com/dotnet/pinvoke/pull/535 if you care to review it.

AArnott commented on src/NerdBank.GitVersioning/Managed/FileHelpers.cs:
```diff
@@ -0,0 +1,101 @@
+        {
+            if (IsWindows)
+            {
+                var handle = CreateFile(path, FileAccess.Read, FileShare.Read, IntPtr.Zero, FileMode.Open, attributes, IntPtr.Zero);
```
Why do we need to p/invoke to get read access to the file? Is it to avoid a file exists check and open separately while still avoiding a first chance exception? If so, that may be reasonable, although a shame that .NET doesn't offer a better alternative in the BCL.

qmfrederik replied:
Yes, that's the reason. You need a file exists check because the file may not exist (for example, when the object is stored in a packfile or an alternate instead).

`File.Exists` came up in the performance traces, so that's why I ended up using the P/Invoke. (Plus, it's racy too).

I don't think the BCL has `File.TryOpen` at the moment; there's an API proposal (dotnet/runtime#27217) though.

AArnott commented on src/NerdBank.GitVersioning/Managed/GitPackReader.cs:
```diff
@@ -0,0 +1,115 @@
+    {
+        private static readonly byte[] Signature = GitRepository.Encoding.GetBytes("PACK");
+
+        public static Stream GetObject(GitPack pack, Stream stream, int offset, string objectType, GitPackObjectType packObjectType)
```
General comment: I've added C# 8 nullable ref annotations to most of my libraries and love it. It's helped me find and avoid many null bugs already. What do you think of bumping `LangVersion=8` for this repo and turning on `#nullable enable` for each of these new files so that you get the benefit right away as you're developing, and to avoid a bunch of engineering debt later when I go through and update everything? You know this area better than I do, so I'd be more confident in your doing it than myself.

qmfrederik replied:
Thanks, I added a first round of nullable annotations. Is there specific guidance for methods like `bool File.TryOpen(string path, out FileStream stream)`? Should the `stream` parameter be annotated nullable (it's only `null` when the method returns `false`)?

AArnott replied:
Thanks for that. I've pushed 29bc0f4 to your branch with the fix for that particular method so you can see how it's done.

AArnott reviewed (changes requested):
This looks exciting. Thanks so much for your work. The perf improvements look significant and the code quality looks really good too. I haven't reviewed everything, but I wanted to share what I have so far with you and get your thoughts before I continue.

qmfrederik commented:
@AArnott Thanks for the feedback! I'll try to get the framework targeting sorted out first, probably early next week.

qmfrederik commented on src/Nerdbank.GitVersioning.Tasks/Nerdbank.GitVersioning.Tasks.csproj:
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp2.0;net461</TargetFrameworks>
+    <TargetFrameworks>netcoreapp3.0</TargetFrameworks>
```
I ended up targeting `netstandard2.0;netcoreapp3.1` in `Nerdbank.GitVersioning` and `net461;netcoreapp2.1` in `NerdBank.GitVersioning.Tasks`, to keep up with the `MSBuildFull` and `MSBuildCore` structure in the NuGet package.

That means that the `netcoreapp3.1`-variant of the code (which is the most optimized, since it can use APIs (like `Stream.Read(Span<byte>)`) which are available starting with `netcoreapp3.1 `) is not currently deployed.

qmfrederik commented on src/Cake.GitVersioning/Cake.GitVersioning.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
```
I don't know much about Cake, so I reverted back to the original value.

qmfrederik commented on src/nbgv/nbgv.csproj:
```diff
@@ -4,7 +4,7 @@
     <PackAsTool>True</PackAsTool>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
```
Done.

qmfrederik commented on src/NerdBank.GitVersioning.Tests/NerdBank.GitVersioning.Tests.csproj:
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>net461;netcoreapp3.1</TargetFrameworks>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
I reverted to `net461;netcoreapp3.1`.

qmfrederik commented:
OK - CI is green, I'll try to address more comments tomorrow.

DomZZ commented:
Whoa, amazing stuff, thanks for all of this.
Don"t forget unchecked things in your TODO list (1st post) please.

I was waiting for a managed GIT implementation a long time ago :)

filipnavara commented:
> Since most commits only have one parent, investigate whether it is worth storing the first parent as a field on the GitCommit struct, and additional parents in a list. This would avoid allocating many List<GitObjectId> objects with only a single item.

Actually the common case is one or two parents, ie. single commit and merge commit. At least that's the experience gained with go-git and gitea on various repositories. The odd case is octo-merge (> 2 parents).

filipnavara commented on src/NerdBank.GitVersioning/Managed/FileHelpers.cs:
```diff
@@ -0,0 +1,125 @@
+            }
+        }
+
+        public static unsafe bool TryOpen(Span<byte> path, CreateFileFlags attributes, out FileStream? stream)
```
Span -> ReadOnlySpan ?

qmfrederik replied:
Yes, thanks!

filipnavara commented on src/NerdBank.GitVersioning/Managed/GitObjectId.cs:
```diff
@@ -0,0 +1,259 @@
+        {
+            fixed (byte* thisValue = this.value)
+            {
+                return new Span<byte>(thisValue, NativeSize).SequenceEqual(new Span<byte>(other.value, NativeSize));
```
Span -> ReadOnlySpan ?

qmfrederik replied:
Yes, thanks!

qmfrederik commented on src/NerdBank.GitVersioning.Benchmarks/GetVersionBenchmarks.cs:
```diff
@@ -0,0 +1,44 @@
+            "NerdBank.GitVersioning;version.json")]
+        public string TestData;
+
+        public string RepositoryName => this.TestData.Split(';')[0];
```
I searched GitHub for repos which use NerdBank.GitVersioning and have non-trivial version history. These came up. Submodules would probably work, I'll add a build leg for the perf test soon & let's see what works best.

AArnott commented:
@qmfrederik Since I'm pushing to you branch as well, if you want to keep using force push, I suggest `--force-with-lease` so you don't inadvertently blow away a commit I've added. :)

qmfrederik commented:
> I suggest `--force-with-lease` so you don't inadvertently blow away a commit I've added.

Cool, learned something new today ;-). Sure, I'll pull/push instead.

qmfrederik commented on src/NerdBank.GitVersioning.Benchmarks/NerdBank.GitVersioning.Benchmarks.csproj:
```diff
@@ -0,0 +1,26 @@
+    <DebugType>portable</DebugType>
+    <DebugSymbols>true</DebugSymbols>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
```
This should be resolved now, I removed most of the properties. (They came from the BenchmarkDotNet template, fyi)

qmfrederik commented on src/NerdBank.GitVersioning.Benchmarks/NerdBank.GitVersioning.Benchmarks.csproj:
```diff
@@ -0,0 +1,26 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Benchmarks now run on `netcoreapp2.1`, `netcoreapp3.1` and `net461` (Windows only).

AArnott replied:
Great. Can you revise this so that the TargetFrameworks is constant (so we build net461 off-Windows too) but that you only run the tests on the .NET Core runtime? Similar to what I did in aa4b94c79c1d.
We have had requests periodically that nb.gv work on mono, but we never could fully support that because of libgit2. By moving to managed git, we should have a better story there. So I'd prefer to get as much building on linux/mac as possible (even net461 targets) so that we can eventually enable running these on mono as well.

qmfrederik commented on src/NerdBank.GitVersioning.Benchmarks/GetVersionBenchmarks.cs:
```diff
@@ -0,0 +1,44 @@
+
+        public string RepositoryName => this.TestData.Split(';')[0];
+
+        public string VersionPath => this.TestData.Split(';')[1];
```
Correct, I've removed it.

qmfrederik commented on src/NerdBank.GitVersioning.Benchmarks/GetVersionBenchmarks.cs:
```diff
@@ -0,0 +1,44 @@
+                this.RepositoryName);
+
+        public Version Version
+        { get; set; }
```
Yup, done.

qmfrederik commented:
@AArnott I think I've addressed most of your comments, with the exception of the usage of PInvoke.Kernel32, which should come soon. CI is green at this moment.

I've added benchmarks to the Azure pipeline. They seem to indicate the managed implementation matches or exceeds the performance of native libgit2. Perf has regressed a bit since the start of this PR, since missing features were added (such as support for including/excluding paths, and fetching the time at which the source code was last modified). I'm sure perf can be improved but is probably already good enough.

The unit tests for `GetExtensions.GitIdAsVersion` and `GitExtensions.GetVersionHeight` don't cover the managed implementation yet; when we finish #527 and the follow-up PR to move these tests to `VersionOracleTests` instead, they should cover the managed implementation too.

AArnott commented:
Great stuff, @qmfrederik. You've done a ton of work. Is all the source code you contributed fresh from you or did some of it get copied from someplace else?

@DomZZ asked:

> Why don't you make an external nuget package for this ? I mean many people just want to manage GIT repo.

What do you think of that at this point, @qmfrederik? Do you want to take on maintenance of a general purpose managed git library or would you rather keep it here and focus on just those scenarios required by NB.GV?

@filipnavara said:

> Actually the common case is one or two parents, ie. single commit and merge commit. At least that's the experience gained with go-git and gitea on various repositories. The odd case is octo-merge (> 2 parents).

I agree, one or two parents would be the by-far majority. I imagine we could create two new fields (`FirstParent` and `SecondParent`) that handle the common case and have a 3rd field for a `List` of additional parents when they apply. We could still have a single `Parents` property that is enumerable for easy consumption. The enumerator could be a custom struct that visits first and second parent fields first and proceeds on to the List if it is non-null. 
@qmfrederik as you've been studying the perf traces, do you think avoiding the allocation of the List and its underlying array would be worthwhile here?

qmfrederik commented:
> Is all the source code you contributed fresh from you or did some of it get copied from someplace else?

This is pretty much a fresh implementation, as the existing implementations were old and not very perf-oriented. The work was originally done here https://github.com/qmfrederik/Quamotion.GitVersioning, with the help of @filipnavara .

A couple of helper methods (`GitObjectId.CopyToUtf16String`, `GitObjectId.CreateString` and potentially `StreamExtensions.ReadMbsInt` and `GitPackReader.ReadVariableLengthInteger`) were inspired by what was previously available, but that's about it.

> Do you want to take on maintenance of a general purpose managed git library

No, I don't have the bandwidth for that.

> We could still have a single Parents property that is enumerable for easy consumption. The enumerator could be a custom struct that visits first and second parent fields first and proceeds on to the List if it is non-null.

It could, though https://github.com/dotnet/Nerdbank.GitVersioning/pull/521/files#diff-6f6218b6aaed3d5871d8211df758572dbf154d455164cf8ac93db1ebc77c9420R176-R203 is the only code which actually iterates over parents. Special-casing would avoid allocating the enumerator.

> As you've been studying the perf traces, do you think avoiding the allocation of the List and its underlying array would be worthwhile here?

When I last ran the traces, this List was what came up on top w.r.t. allocations. I would have to re-run the traces now; 64f6c4552ceb4560d463a6d05b21d4e29f181c2c added allocations of at least two strings for every commit, which I believe is unnecessary.

AArnott commented:
> Special-casing would avoid allocating the enumerator.

No enumerator is allocated. Because you typed it as `List<T>` (the concrete type) the compiler can use `List<T>.Enumerator` (the struct) as a struct without boxing. So enumerating should be alloc-free. So the optimization is really about avoiding allocating the `T[]` and `List<T>` objects.

> When I last ran the traces, this List was what came up on top w.r.t. allocations. 

Awesome. Are you interested in making that optimization then? Otherwise I might find time sometime tomorrow.

> I would have to re-run the traces now; 64f6c45 added allocations of at least two strings for every commit, which I believe is unnecessary.

I agree: NB.GV has no use for author info.

qmfrederik commented:
> Awesome. Are you interested in making that optimization then? Otherwise I might find time sometime tomorrow.

I'd like to work on the unit test coverage first; so if you want to have a go at it, please do! Never hurts to have another set of eyes look at the same problem.

Would you mind having a look at #527, though? It'll enable me to extend unit test coverage here.

> I agree: NB.GV has no use for author info.

Except for #329 ;-)

qmfrederik commented on src/NerdBank.GitVersioning/Managed/FileHelpers.cs:
```diff
@@ -0,0 +1,104 @@
+
+                if (!handle.IsInvalid)
+                {
+                    var fileHandle = new SafeFileHandle(handle.DangerousGetHandle(), ownsHandle: true);
+                    handle.SetHandleAsInvalid();
```
We can simplify this if `PInvoke.Kernel32.CreateFile` returns a `SafeFileHandle` instead of a `SafeObjectHandle`.

AArnott replied:
That would be a breaking API change at this point. I don't recall why we went with that API at the time.

qmfrederik commented:
macOS build failure (failure to provision a VM) seems unrelated.

qmfrederik commented:
So, a couple of perf optimizations managed to increase the performance even further
- Accessing the Git pack index as a memory mapped file makes it really cheap to determine whether an object exists in the git pack or not (it becomes an in-memory binary search)
- Accessing the Git packfile as a memory mapped file allows us to open the packfile only once (i.e. single call to `CreateFile`, which is _very_ expensive); all `Stream` objects which represents the (compressed) individual objects can read from the memory mapped file
- Having `GitRepository` search pack files _before_  looking for loose files significantly improves performance (mainly because `CreateFile` is expensive, even if the file does not exist and the method just returns an error)

qmfrederik commented:
|            Method |           Job |       Runtime |     ProjectDirectory |      Mean |     Error |    StdDev |    Median | Ratio | RatioSD |
|------------------ |-------------- |-------------- |--------------------- |----------:|----------:|----------:|----------:|------:|--------:|
| GetVersionLibGit2 |    .NET 4.6.1 |    .NET 4.6.1 |               Cuemon | 45.156 ms | 2.0427 ms | 5.8608 ms | 44.679 ms |  1.13 |    0.22 |
| GetVersionManaged |    .NET 4.6.1 |    .NET 4.6.1 |               Cuemon | 31.963 ms | 2.0405 ms | 5.9844 ms | 30.600 ms |  0.81 |    0.21 |
| GetVersionLibGit2 | .NET Core 2.1 | .NET Core 2.1 |               Cuemon | 42.214 ms | 2.2380 ms | 6.5283 ms | 40.579 ms |  1.05 |    0.22 |
| GetVersionManaged | .NET Core 2.1 | .NET Core 2.1 |               Cuemon | 11.517 ms | 0.8661 ms | 2.4710 ms | 11.226 ms |  0.29 |    0.08 |
| GetVersionLibGit2 | .NET Core 3.1 | .NET Core 3.1 |               Cuemon | 40.857 ms | 2.2597 ms | 6.4471 ms | 39.805 ms |  1.00 |    0.00 |
| GetVersionManaged | .NET Core 3.1 | .NET Core 3.1 |               Cuemon | 10.160 ms | 0.4483 ms | 1.3148 ms |  9.834 ms |  0.26 |    0.05 |
|                   |               |               |                      |           |           |           |           |       |         |
| GetVersionLibGit2 |    .NET 4.6.1 |    .NET 4.6.1 | NerdB(...)oning [22] |  6.843 ms | 0.5259 ms | 1.5089 ms |  6.328 ms |  0.98 |    0.24 |
| GetVersionManaged |    .NET 4.6.1 |    .NET 4.6.1 | NerdB(...)oning [22] |  3.807 ms | 0.4315 ms | 1.1884 ms |  3.410 ms |  0.55 |    0.20 |
| GetVersionLibGit2 | .NET Core 2.1 | .NET Core 2.1 | NerdB(...)oning [22] |  5.944 ms | 0.3051 ms | 0.8804 ms |  5.773 ms |  0.85 |    0.15 |
| GetVersionManaged | .NET Core 2.1 | .NET Core 2.1 | NerdB(...)oning [22] |  2.375 ms | 0.1151 ms | 0.3227 ms |  2.246 ms |  0.34 |    0.07 |
| GetVersionLibGit2 | .NET Core 3.1 | .NET Core 3.1 | NerdB(...)oning [22] |  7.004 ms | 0.3054 ms | 0.8958 ms |  6.842 ms |  1.00 |    0.00 |
| GetVersionManaged | .NET Core 3.1 | .NET Core 3.1 | NerdB(...)oning [22] |  2.404 ms | 0.0478 ms | 0.1108 ms |  2.390 ms |  0.34 |    0.04 |
|                   |               |               |                      |           |           |           |           |       |         |
| GetVersionLibGit2 |    .NET 4.6.1 |    .NET 4.6.1 |          SuperSocket | 19.695 ms | 0.7297 ms | 2.0581 ms | 19.458 ms |  0.83 |    0.22 |
| GetVersionManaged |    .NET 4.6.1 |    .NET 4.6.1 |          SuperSocket | 11.717 ms | 1.1912 ms | 3.4747 ms | 10.318 ms |  0.50 |    0.19 |
| GetVersionLibGit2 | .NET Core 2.1 | .NET Core 2.1 |          SuperSocket | 23.583 ms | 1.6592 ms | 4.8401 ms | 22.716 ms |  1.01 |    0.32 |
| GetVersionManaged | .NET Core 2.1 | .NET Core 2.1 |          SuperSocket |  5.673 ms | 0.3592 ms | 0.9892 ms |  5.484 ms |  0.24 |    0.07 |
| GetVersionLibGit2 | .NET Core 3.1 | .NET Core 3.1 |          SuperSocket | 24.840 ms | 2.1175 ms | 6.1769 ms | 23.024 ms |  1.00 |    0.00 |
| GetVersionManaged | .NET Core 3.1 | .NET Core 3.1 |          SuperSocket |  6.359 ms | 0.4603 ms | 1.2982 ms |  6.116 ms |  0.27 |    0.08 |
|                   |               |               |                      |           |           |           |           |       |         |
| GetVersionLibGit2 |    .NET 4.6.1 |    .NET 4.6.1 |                xunit | 19.834 ms | 0.8643 ms | 2.4658 ms | 19.308 ms |  0.89 |    0.24 |
| GetVersionManaged |    .NET 4.6.1 |    .NET 4.6.1 |                xunit |  9.714 ms | 0.4631 ms | 1.3136 ms |  9.360 ms |  0.44 |    0.12 |
| GetVersionLibGit2 | .NET Core 2.1 | .NET Core 2.1 |                xunit | 22.281 ms | 1.7284 ms | 4.9591 ms | 21.236 ms |  1.00 |    0.32 |
| GetVersionManaged | .NET Core 2.1 | .NET Core 2.1 |                xunit |  7.139 ms | 0.6885 ms | 1.9644 ms |  6.583 ms |  0.32 |    0.12 |
| GetVersionLibGit2 | .NET Core 3.1 | .NET Core 3.1 |                xunit | 23.684 ms | 2.0372 ms | 5.9746 ms | 22.271 ms |  1.00 |    0.00 |
| GetVersionManaged | .NET Core 3.1 | .NET Core 3.1 |                xunit |  6.015 ms | 0.5195 ms | 1.4481 ms |  5.525 ms |  0.27 |    0.08 |

qmfrederik commented:
macOS build failure (failure to provision a VM) seems unrelated (again)

AArnott commented:
> I'd like to work on the unit test coverage first; so if you want to have a go at it, please do! Never hurts to have another set of eyes look at the same problem.

I pushed the `List<GitCommitId>` optimization.

qmfrederik commented:
> I pushed the List<GitCommitId> optimization.

Awesome! Seems to have pushed the needle even further by a couple of percentage points. On the other hand, 
0c5a573, which I pushed, caused a significant regression so I reverted it.

I merged master and fixed a couple of issues. I'll try to light up the GetVersionHeight unit tests.

qmfrederik commented:
@AArnott So - the managed `VersionOracle` originated from a lean & mean prototype I had, but I was unable to find a straightforward way to add all of the subtle and not-so-subtle NBGV features.

I ended up with a _new_ implementation of the `ManagedVersionOracle` which is a copy & paste of the `LibGit2VersionOracle`, but uses the managed backend instead of LibGit2.

Perf regressed a bit but is still ahead of the LibGit2-based implementation and more importantly, unit tests pass.

Is it OK for you to finish up this PR like that? It achieves the stated goal - a managed implementation so we no longer depend on LibGit2 (which means VS Code support, important to me), maintains correctness and pushes perf.

AArnott commented:
Thanks for the update. I'll do a comprehensive review + touch up pass on this in the next 48 hours with the goal to merge.

AArnott commented:
The testing in this is now quite patchy. And it's not clear looking at a test whether it will exercise the managed or libgit2 implementation or even both in some cases. I'm doing some rearranging of code to make it much more clear, and see if I can get the full test coverage over both implementations.

AArnott commented:
I think for adequate testing and control of what code we're calling, and to enable more code sharing as a bonus, I'm going to change `VersionFile` and some of the 'extension' methods into a factory pattern so code is running entirely in one world or the other. We definitely had a lot of code in tests, and some in the product, that was using a mixture. 
Once we have the factory pattern in place, I can adjust the tests to run against both instead of just one or an undefined mix.

qmfrederik commented:
> I think for adequate testing and control of what code we're calling, and to enable more code sharing as a bonus, I'm going to change VersionFile and some of the 'extension' methods into a factory pattern so code is running entirely in one world or the other.

Awesome, thanks!

AArnott commented:
What I'm working on will change around a lot of code. Please do *not* make changes in the meantime to avoid merge conflicts.

qmfrederik commented:
@AArnott I did push one commit which should add better support for large packfiles (I encountered a large packfile which I could use to run some tests against). I assume it's orthogonal to what you're working on so hopefully no merge conflicts there.

AArnott commented:
Sounds good. I haven't touched the `ManagedGit` folder at all, so that should be fine.
I haven't forgotten about this PR. I have a huge set of changes that I've been working on periodically as I have time. My API changes are done, and I'm fixing up tests to compile. I have 85 errors left to go. I'm sure there will be a lot of failures for me to work through, but the good part is when it's done, all the tests will run against both implementations. The msbuild task will similarly be able to run against both backends (and default to the managed one, of course) so we can adequately test both and allow folks to switch back to libgit2 if the managed implementation ends up missing a scenario in the short term.

AArnott commented:
@qmfrederik, I'm _very_ close. I have all the libgit2 tests passing and almost all the tests passing when they're run against the managed implementation. But one test in particular (`VersionJson_InheritButNoParentFileFound`) is failing and while I investigated I saw some very odd behavior. When I run it in the debugger and stop at `TrySelectCommit("HEAD")`, the commit it loads has an ID where only the first byte (two hex characters) is initialized while the rest are 0s. But when I look at the underlying bytes, they are all initialized. I think `ParseHex` is misbehaving. Any ideas? You'll have to check my branch on this root repo to see it.

AArnott commented:
Oh, never mind. I guess the fixed size buffer in the struct is just fooling the debugger into only showing the first element in the array.

AArnott commented:
I have all tests passing. Just a few compiler warnings to clean up.

AArnott commented:
So @qmfrederik  where are we at, would you say? I see a few checkboxes still clear, but didn't you already look into large pack files? I can help test the shallow clone scenario. Did you want to look into parallel tree walking, or should we save that for latter or just forget it, given the perf wins you already made?

qmfrederik commented:
> When I run it in the debugger and stop at TrySelectCommit("HEAD"), the commit it loads has an ID where only the first byte (two hex characters) is initialized while the rest are 0s.

Yes, I've seen this as well. I have no idea why, but it appears to be an issue with the debugger (as you've pointed out). There appears to be no runtime impact. But it's very odd indeed.

> But when I look at the underlying bytes, they are all initialized. I think ParseHex is misbehaving.

Now that .NET 5.0 has been released, we may consider using `HexConvert` (dotnet/runtime#17837), which is being optimized (dotnet/runtime#37546) on that platform.

qmfrederik commented:
> but didn't you already look into large pack files?

Yes, I ran some tests on large packfiles. It resulted in the method which operate on packfiles taking `long`s as offsets (78f27aab92b58dc631daab6b8dee9053fc17c03c).

> I can help test the shallow clone scenario.

Cool, thanks!.

> Did you want to look into parallel tree walking, or should we save that for latter or just forget it, given the perf wins you already made?

Yeah, I don't think it really is a priority right now. So I'd say it is a candidate for a separate PR at a later point in time, if needed.

qmfrederik commented:
> The msbuild task will similarly be able to run against both backends (and default to the managed one, of course) so we can adequately test both and allow folks to switch back to libgit2 if the managed implementation ends up missing a scenario in the short term.

This is awesome. I'm sure there'll be things we missed, so being able have a fallback scenario for now seems like a good thing.

AArnott commented:
@qmfrederik Do you have time to work on the last couple of items on the checklist?
I'm writing tests for `GitContext` now that will include testing this behavior, and it will surely fail at this point.

AArnott commented:
@qmfrederik I've been reading up on the .git directory's format for refs and I think I can implement at least part of the support we need for resolving a ref to a commit. I'm going to give it a try unless you tell me you're already working on it.

AArnott commented:
Ok, we just have one failing test (`SelectCommitByPartialId`) that will require searching in the objects directory for loose files as well as the pack file. We'll presumably need to detect ambiguous partial commits even when both the pack file and the directory structure have exactly one match (so checking both locations is always important).
As I have no background in the pack file format or the API you created, can you fix this last test, @qmfrederik ?

qmfrederik commented:
Sure, @AArnott . I implemented Lookup, which searches both the objects directory as well as the pack files. All unit tests are now green.

AArnott commented:
Awesome. Then this may be ready to merge. There is that last regression that in the managed git world we don't implement finding the shortest possible commit id (`GitRepository.ShortenObjectId`) which is used to produce an optimally sized +buildMetadata part of the version. I'd like to keep that functionality since we expose it in `version.json`. 
@qmfrederik do you think implementing that is cheap enough that you can find time to do so?

AArnott commented:
@qmfrederik I'm also wondering where you learned about the internals of git pack file formats. [The Git Book doesn't go](https://git-scm.com/book/en/v2/Git-Internals-Packfiles) into that kind of detail. 
I'm thinking about a while from now where perhaps your interests have shifted and I have to maintain this and where I can go to learn about it.

qmfrederik commented:
@AArnott, re: Git Pack documentation, here are the documents I found useful:

- https://git-scm.com/docs/pack-format describes the file format
- https://stefan.saasen.me/articles/git-clone-in-haskell-from-the-bottom-up/#format-of-the-delta-representation goes in further detail, with examples

AArnott commented:
Thanks for that. I did write a test for the short commit id, but it's based on libgit2's (odd?) behavior of not returning anything less than 7 as a length. It's in the `features/managed-git` branch of *this* repo.
@qmfrederik If you don't want to spend the time filling in this last gap, I'll probably merge this PR and open an issue to track it and I'll try to fill it in before shipping a stable update. Just let me know, as you've been extremely helpful so far and I don't want to abuse that. :)

qmfrederik commented:
@AArnott Well, I did a quick implementation of `ShortenObjectId` which starts with `minimum`, checks whether `Lookup(...)` returns non-`null` (in which case it is well-defined), and then increments the length until the maximum length of 40 is received.

I guess libgit2 does always returns 7 characters to make sure the values are somehow future-proof.

AArnott commented:
Great. I'm not concerned about perf for partial commit id matching since the user scenario is a `nbgv get-version partial-commit` invocation, so a few loops around to look for matches won't be perceivable to the user.

However, in adding a couple tests I found that odd-number length partial commits produces an exception. I guess since you're converting the hex to bytes it makes searching by bytes hard when you only have half a byte prescribed. Any ideas? I wouldn't want to cause builds to fail when they specified `7` or `11` as their target commit length.

AArnott commented:
I just added a test for odd length commit lookups by partial id and it threw the same exception.

AArnott commented:
I'm trying my hand at a fix for this, unless you'd rather, @qmfrederik

AArnott commented:
Got it fixed. :) Merging tonight...

AArnott reviewed (approved):
Your contribution @qmfrederik is amazing. Thank you so much!

AArnott commented:
And thank you @filipnavara for helping with the awesome performance improvements too!

AArnott commented:
If all goes well, this will push to nuget.org as 3.4.142-alpha in about 30 minutes.

qmfrederik commented:
Thanks, @AArnott ! It looks like the package didn't make it to nuget.org, but it's available in the CI feed, so that works for me :).

AArnott commented:
Ya, it got snagged due to a bad service connection. That has been rectified though so it's on nuget.org now.
